File: longest-substring-with-at-least-k-repeating-characters.java
// https://leetcode.com/problems/longest-substring-with-at-least-k-repeating-characters/description/
// Count & Bit-masking
// Time: O(n^2), Space: O(1)
class Solution {
    public int longestSubstring(String s, int k) {
        int maxLength = 0;
    	for(int i=0, maxJ=0; i<=s.length()-k; i=++maxJ) {
    		int atozCount[] = new int[26];
    		int atozMask = 0;
    		for(int j=i; j<s.length(); j++) {
    			int index = s.charAt(j) - 'a';
    			int count = ++atozCount[index];
    			if(count<k) atozMask |= 1<<index;
    			else atozMask &= ~(1<<index);
    			
    			if(atozMask==0) {
    				maxLength = Math.max(maxLength, j-i+1);
    				maxJ = j;
    			}
    		}
    	}
        return maxLength;
    }
}

File: longest-substring-with-at-least-k-repeating-characters.py
# https://leetcode.com/problems/longest-substring-with-at-least-k-repeating-characters/
# Divide & Conquer
# Time: O(n^2), Space: O(n^3)
class Solution:
    def longestSubstring(self, s, k):
        #print(s)
        if len(s)<k: return 0
        
        cnts = [0 for x in range(26)]
        for c in s: cnts[ord(c)-97] += 1
        
        isAllAns, start, maxLen = True, 0, 0
        for i in range(len(s)):
            if cnts[ord(s[i])-97]<k:
                isAllAns = False
                maxLen = max(maxLen, self.longestSubstring(s[start:i],k))
                start = i+1
        
        if isAllAns: return len(s)
        else: return max(maxLen, self.longestSubstring(s[start:],k))
